### PR TITLE
Ea 2758 content stats using solr

### PR DESCRIPTION
**Pull request by SrdjanStevanetic**

### CODE REVIEW

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+public class AnnotationStatisticsService {
+
+	@Resource
+	SolrAnnotationService solrService;
+
+    public void getAnnotationsStatistics(AnnotationMetric annoMetric) throws AnnotationServiceException {
+    	annoMetric.setTimestamp(new Date());  
+    	//getting the annotations statistics for the clients
+    	QueryResponse annoFacetStatsJson = solrService.getAnnotationStatisticsJsonNestedFacets(SolrAnnotationConstants.GENERATOR_URI);
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:30]: there is an open question in the ticket to clear out which is the main facet, the scenario or the generator (client application)

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    	annoFacetStatsJson = solrService.getAnnotationStatisticsJsonFacets(); 
+    	extractAnnotationStatisticsOverallFromJsonFacet(annoFacetStatsJson, annoMetric);
+    }
+    
+    private void extractAnnotationStatisticsClientUserFromJsonNestedFacet (QueryResponse annoFacetStats, AnnotationMetric annoMetric, String facetField, String target) {
+    	List<AnnotationStatistics> annotationStatistics = new ArrayList<AnnotationStatistics>();
+	    annoFacetStats.getJsonFacetingResponse().getBucketBasedFacets(facetField).getBuckets();
+    	for (BucketJsonFacet mainBucket : annoFacetStats.getJsonFacetingResponse().getBucketBasedFacets(facetField).getBuckets()) {
+    		AnnotationStatistics annoStats = null;
+    		if(target.compareToIgnoreCase(AnnotationStatisticsConstants.CLIENT)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:45]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    			annoStats = new AnnotationStatisticsClient();
+    			((AnnotationStatisticsClient)annoStats).setClient(mainBucket.getVal().toString());
+    		}
+    		else if(target.compareToIgnoreCase(AnnotationStatisticsConstants.USER)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:49]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    			((AnnotationStatisticsUser)annoStats).setUser(mainBucket.getVal().toString());
+    		}
+    		
+    		if (annoStats==null) continue;
+		    	
+    		setAnnotationStatisticsFromJsonBuckets(annoStats, mainBucket.getBucketBasedFacets(SolrAnnotationConstants.SCENARIO).getBuckets());
+        	annotationStatistics.add(annoStats);
+    	}
+	    
+		if(target.compareToIgnoreCase(AnnotationStatisticsConstants.CLIENT)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:60]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+			annoMetric.setAnnotationStatisticsClients(annotationStatistics);
+		}
+		else if(target.compareToIgnoreCase(AnnotationStatisticsConstants.USER)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:63]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    
+    private void extractAnnotationStatisticsOverallFromJsonFacet (QueryResponse annoFacetStats, AnnotationMetric annoMetric) {
+    	AnnotationStatistics annotationStatisticsOverall = new AnnotationStatistics();	
+    	setAnnotationStatisticsFromJsonBuckets(annotationStatisticsOverall,annoFacetStats.getJsonFacetingResponse().getBucketBasedFacets(SolrAnnotationConstants.SCENARIO).getBuckets());
+    	annoMetric.setAnnotationStatisticsScenarios(annotationStatisticsOverall);
+    }
+    
+    private void setAnnotationStatisticsFromJsonBuckets (AnnotationStatistics annoStats, List<BucketJsonFacet> facetJsonBuckets) {
+    	for (BucketJsonFacet facetJsonBucket : facetJsonBuckets) {  
+        	if(facetJsonBucket.getVal().toString().compareToIgnoreCase(AnnotationScenarioTypes.GEO_TAG)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:76]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        		annoStats.setGeoTag(facetJsonBucket.getCount());
+        	}
+        	if(facetJsonBucket.getVal().toString().compareToIgnoreCase(AnnotationScenarioTypes.TRANSCRIPTION)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:79]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        		annoStats.setTranscription(facetJsonBucket.getCount());
+        	}
+        	if(facetJsonBucket.getVal().toString().compareToIgnoreCase(AnnotationScenarioTypes.OBJECT_LINK)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:82]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        		annoStats.setObjectLink(facetJsonBucket.getCount());
+        	}
+        	if(facetJsonBucket.getVal().toString().compareToIgnoreCase(AnnotationScenarioTypes.SEMANTIC_TAG)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:85]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        		annoStats.setSemanticTag(facetJsonBucket.getCount());
+        	}
+        	if(facetJsonBucket.getVal().toString().compareToIgnoreCase(AnnotationScenarioTypes.SUBTITLE)==0) {
```
Comment (gsergiu) [inline annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:88]: please switch the position of the constant and use equalsIgnoreCase

File: annotation-solr/src/test/resources/solr7-dev/anno-dev/anno-up/conf/schema.xml [hunk 11]
```diff
@@ -145,7 +150,7 @@
 			
 		<!-- @Field("body_value.*") NEW -->
 		<dynamicField name="body_value.*" type="text" required="false"
-			multiValued="false" indexed="true" stored="true" />	
+			multiValued="true" indexed="true" stored="true" />
```
Comment (gsergiu) [inline annotation-solr/src/test/resources/solr7-dev/anno-dev/anno-up/conf/schema.xml:153]: we don't have multiple values for individual languages.  This should stay with multivalues=false

**Review by gsergiu — CHANGES_REQUESTED** [review]

File: annotation-statistics/pom.xml [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+				</includes>
+			</resource>
+		</resources>
+		<testResources>
+			<testResource>
+				<directory>src/test/resources</directory>
+			</testResource>
+		</testResources>
+
+		<plugins>
```
[the remark this answers is not in the document]
Comment (SrdjanStevanetic) [reply annotation-statistics/pom.xml:75]: Makes sense.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/model/AnnotationMetric.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import eu.europeana.annotation.statistics.vocabulary.AnnotationStatisticsConstants;
+
+@JsonInclude(value = JsonInclude.Include.NON_EMPTY)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AnnotationMetric {
```
[the remark this answers is not in the document]
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/model/AnnotationMetric.java:14]: Here I kept the class without inheriting. The reason is that the Metric class is not available in the current version of the api-commons we use in the annotation-api, which is 0.3.7, and if we switch to 0.3.11 which has the Metric class, that would trigger several dependency changes. First the new dependency to the corelib-definitions needs to be introduced, then the class BaseAddress needs to be changed to the AddressImpl from the corelib-storage module. I was not sure if we want it that way, so I kept the current api-commons version without inheriting the Metric class. Anyway that class has only the type and the timestamp, which both are not required for the output. Btw, the AnnotationMetic class I would keep in the annotation-api module and not move to the api commons, becuase it is not shared with any other module.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/model/AnnotationStatisticsClient.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import eu.europeana.annotation.statistics.vocabulary.AnnotationStatisticsConstants;
+
+@JsonInclude(value = JsonInclude.Include.NON_EMPTY)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AnnotationStatisticsClient extends AnnotationStatistics {
```
[the remark this answers is not in the document]
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/model/AnnotationStatisticsClient.java:11]: The same comment as the above one.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/model/AnnotationStatisticsUser.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import eu.europeana.annotation.statistics.vocabulary.AnnotationStatisticsConstants;
+
+@JsonInclude(value = JsonInclude.Include.NON_EMPTY)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AnnotationStatisticsUser extends AnnotationStatistics {
```
[the remark this answers is not in the document]
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/model/AnnotationStatisticsUser.java:11]: The same comment as the above one.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 1 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:30]: I kept it the way it is defined in the specs.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 2 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:45]: This code is not needed any more.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 3 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:49]: This code is not needed any more.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 4 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:60]: This code is not needed any more.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 5 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:63]: This code is not needed any more.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 6 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:76]: This code is not needed any more.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 7 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:79]: This code is not needed any more.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 8 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:82]: This code is not needed any more.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 6 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:76]: I do not know what you mean by switching the position of the constant, that the constant is the first liek this: 
if(AnnotationScenarioTypes.GEO_TAG.equalsIgnoreCase(facetJsonBucket.getVal().toString()). For me the way I wrote is a bit more natural. You can correct it the way you want.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 7 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:79]: Improved.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 8 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:82]: Improved.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 9 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:85]: Improved.

File: annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java (hunk 10 above)
Comment (SrdjanStevanetic) [reply annotation-statistics/src/main/java/eu/europeana/annotation/statistics/service/AnnotationStatisticsService.java:88]: Improved.

File: annotation-solr/src/test/resources/solr7-dev/anno-dev/anno-up/conf/schema.xml (hunk 11 above)
Comment (SrdjanStevanetic) [reply annotation-solr/src/test/resources/solr7-dev/anno-dev/anno-up/conf/schema.xml:153]: Changed.

**Pull request merged**